File: Src/Icm.Core.Tests/Compilation/CompiledFunctionTest.vb
Imports System.CodeDom.Compiler
Imports Icm.Compilation

<TestFixture(), Category("Icm")>
Public Class CompiledFunctionTest

    Private ReadOnly Namespaces As New List(Of String)()

    <Test()>
    Public Sub CompileOkTest()
        Dim Funct As CompiledFunction(Of Integer)
        Funct = New VBCompiledFunction(Of Integer)
        Funct.AddParameter(Of Integer)("x")
        Funct.AddParameter(Of Integer)("y")
        Funct.Code = "x+y"
        Funct.CompileAsExpression()
        Assert.That(Funct.CompilerErrors.Count = 0)
    End Sub

    <Test()>
    Public Sub EvaluateTest()
        Dim Funct As CompiledFunction(Of Integer)
        Funct = New VBCompiledFunction(Of Integer)
        Funct.AddParameter(Of Integer)("x")
        Funct.AddParameter(Of Integer)("y")
        Funct.Code = "x+y"
        Funct.CompileAsExpression()
        Assert.AreEqual(10, Funct.Evaluate(5, 5))
    End Sub

    <Test()>
    Public Sub CompileSyntaxErrorTest()

        Dim Funct As New VBCompiledFunction(Of Integer)
        Funct.AddParameter(Of Integer)("x")
        Funct.AddParameter(Of Integer)("y")
        Funct.Code = "x..y"

        Assert.Throws(Of CompileException)(Sub()
                                               Funct.CompileAsExpression()
                                           End Sub)
        Dim list = Funct.CompilerErrors
        Dim errores = list.Where(Function(ce) Not ce.IsWarning)
        Dim warnings = list.Where(Function(ce) ce.IsWarning)
        Assert.That(errores.Count = 1)
        Assert.That(warnings.Count = 0)
    End Sub

End Class

File: Src/Icm.Core.Tests/Compilation/CompiledFunctionTest.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.CodeDom.Compiler;
using Icm;
using Icm.Compilation;
using NUnit.Framework;

[TestFixture(), Category("Icm")]
public class CompiledFunctionTest
{

	[TestCase("Nothing", 1, 1, ExpectedResult = 0)]
	[TestCase("Nothing", 2, 5, ExpectedResult = 0)]
	[TestCase("x+y", 1, 1, ExpectedResult = 2)]
	[TestCase("x+y", 2, 5, ExpectedResult = 7)]
	[TestCase("x*y", 1, 1, ExpectedResult = 1)]
	[TestCase("x*y", 2, 5, ExpectedResult = 10)]
	[TestCase("x..y", 1, 1, ExpectedException = typeof(CompileException))]
	public int Evaluate_Test(string code, int param1, int param2)
	{
		CompiledFunction<int> Funct = default(CompiledFunction<int>);
		Funct = new VBCompiledFunction<int>();
		Funct.AddParameter<int>("x");
		Funct.AddParameter<int>("y");
		Funct.Code = code;
		Funct.CompileAsExpression();
		Assert.That(Funct.CompilerErrors.Count == 0);

		return Funct.Evaluate(param1, param2);
	}

}

[TestFixture(), Category("Icm")]
public class VBCompiledFunctionTest
{

	[Test()]
	public void GeneratedCode_Test()
	{
		CompiledFunction<int> Funct = default(CompiledFunction<int>);
		Funct = new VBCompiledFunction<int>();
		Funct.AddParameter<int>("x");
		Funct.AddParameter<int>("y");
		Funct.Code = "x+y*y";

		dynamic expected = "Imports System" + Constants.vbCrLf + "Imports System.Xml" + Constants.vbCrLf + "Imports System.Data" + Constants.vbCrLf + "Imports System.Linq" + Constants.vbCrLf + "Imports Microsoft.VisualBasic" + Constants.vbCrLf + "Namespace dValuate" + Constants.vbCrLf + " Class EvalRunTime " + Constants.vbCrLf + "  Public Function EvaluateIt( _" + Constants.vbCrLf + "   ByVal x As Int32, _" + Constants.vbCrLf + "   ByVal y As Int32 _" + Constants.vbCrLf + "  ) As Int32" + Constants.vbCrLf + "    Return x+y*y" + Constants.vbCrLf + "  End Function" + Constants.vbCrLf + " End Class " + Constants.vbCrLf + "End Namespace" + Constants.vbCrLf;

		Assert.That(Funct.GeneratedCode, Is.EqualTo(expected));
	}


}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
